File: phalcon-mvc/app/library/Console/Application.php
<?php

/*
 * Copyright (C) 2014-2018 The OpenExam Project
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

// 
// File:    Application.php
// Created: 2014-09-08 12:45:04
// 
// Author:  Anders Lövgren (Computing Department at BMC, Uppsala University)
// 

namespace OpenExam\Library\Console;

/**
 * Command option scanner.
 */
class CommandOption
{

        public $key;
        public $val;

        public function __construct($args)
        {
                $match = array();
                if (preg_match("/--(.*?)=(.*)/", $args, $match)) {
                        $this->key = $match[1];
                        $this->val = $match[2];
                } elseif (preg_match("/--(.*)/", $args, $match)) {
                        $this->key = $match[1];
                } elseif ($args[0] == '-') {
                        $this->key = substr($args, 1);
                } elseif (preg_match("/(.*?)=(.*)/", $args, $match)) {
                        $this->key = $match[1];
                        $this->val = $match[2];
                } else {
                        $this->key = $args;
                }
        }

}

/**
 * Console application base class.
 *
 * @author Anders Lövgren (Computing Department at BMC, Uppsala University)
 */
class Application extends \Phalcon\CLI\Console
{

        /**
         * Process command line arguments.
         * 
         * Tasks works similar to controllers: Given the task, action and params
         * argument, the target task/action is called using params as parameters. 
         * 
         * This function process $argument parameter and command line arguments 
         * (from PHP builtin $argv). The merged array is used when calling the
         * parent handle() method.
         * 
         * Both long (--key[=val]) and short (-k) command line options are 
         * supported. Command line options can also be [task [action [params...]]] 
         * 
         * Example 1:
         * <code>
         * // Script database.php:
         * $console = new Application($di);
         * $console->process(array('task' => 'database'));
         * 
         * // Run from command line:
         * php script/database.php --migrate exams students
         * 
         * // The task definition after process():
         * Array
         * (
         *      [task] => database
         *      [action] => migrate
         *      [params] => Array
         *          (
         *              [0] => exams
         *              [1] => students
         *          )
         * )
         * <code>
         * 
         * Example 2 (same task called):
         * <code>
         * // Script cli.php:
         * $console = new Application($di);
         * $console->process();
         * 
         * // Run from command line:
         * php script/cli.php --database --migrate exams students
         * 
         * // The task definition after process():
         * Array
         * (
         *      [task] => database
         *      [action] => migrate
         *      [params] => Array
         *          (
         *              [0] => exams
         *              [1] => students
         *          )
         * )
         * <code>
         * 
         * @param array $arguments The task specification.
         */
        public function process($arguments = null)
        {
                global $argv;
                $options = array();

                foreach ($argv as $args) {
                        $option = new CommandOption($args);
                        if (isset($option->val)) {
                                $options[] = $option->key;
                                $options[] = $option->val;
                        } else {
                                $options[] = $option->key;
                        }
                        unset($option);
                }

                if (strstr($options[0], '.php')) {
                        array_shift($options);
                }

                if (!isset($arguments)) {
                        $arguments = array();
                }

                foreach ($options as $args) {
                        if (!isset($arguments['task'])) {
                                $arguments['task'] = $args;
                        } elseif (!isset($arguments['action'])) {
                                $arguments['action'] = $args;
                        } elseif (!isset($arguments['params'])) {
                                $arguments['params'] = array($args);
                        } else {
                                $arguments['params'][] = $args;
                        }
                }

                parent::handle($arguments);
        }

}

File: admin/simulate/application.cpp
// Copyright (C) 2016  Anders Lövgren, BMC Computing Department, Uppsala University
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// --------------------------------------------------------------------------
// Author: Anders Lövgren <anders.lovgren@bmc.uu.se>
//
// File:   application.cpp
// Date:   4 Feb 2016
// --------------------------------------------------------------------------
//

#include <fstream>
#include <sstream>
#include <unistd.h>

#include "application.hpp"
#include "options.hpp"
#include "output.hpp"

Application::Application(int argc, char **argv)
{
    options = new Options;
    options->Parse(argc, argv);
}

Application::~Application()
{
    delete options;
}

//
// Process current options.
//
void Application::Process()
{
    if(options->HasTaskFile()) {
        Process(options->GetTaskFile());
    } else {
        Process(options->GetTask());
    }
}

//
// Process single task.
//
void Application::Process(const Task &task) const
{
    Process(new Task(task));
}

//
// Process single task.
//
void Application::Process(Task *task) const
{
    task->setTaskObserver(this);
    task->Start();
}

//
// Process task definition file.
//
void Application::Process(const std::string &file)
{
    std::ifstream stream(file, std::ios::in);

    if(!stream) {
        throw Exception("Failed open " + file);
    }

    Process(stream);
}

//
// Process task definition stream.
//
void Application::Process(std::ifstream &stream)
{
    Start(stream);
    Join();
}

//
// Start all task from input stream.
//
void Application::Start(std::ifstream &stream)
{
    std::string line;
    int index = 0;

    const Options::Start &start = options->GetStart();

    while(std::getline(stream, line)) {
        if(Start(line)) {
            if((index != 0) && (index % start.num == 0)) {
                ::sleep(start.wait);
            }
        }
    }
}

//
// Start task defined by input string (read from file). The task is
// merged with common task options before started. The defined task
// options have precedence.
//
bool Application::Start(const std::string &line)
{
    Task *task = new Task(options->GetTask());
    task->Scan(line);

    return Start(task);
}

//
// Start task in own thread. If successful, the task is inserted in the
// list of running tasks.
//
bool Application::Start(Task *task)
{
    Thread thread;

    if(pthread_create(&thread, 0, Task::Start, task) != 0) {
        perror("pthread_create");
        return false;
    }

    tasks[thread] = task;
    task->setTaskObserver(this);

    return true;
}

//
// Join all running threads (tasks).
//
void Application::Join()
{
    for(TaskIterator it = tasks.begin(); it != tasks.end(); ++it) {
        Join(it->first, it->second);
    }
}

//
// Join thread and collect results.
//
void Application::Join(Application::Thread thread, Task *task)
{
    if(pthread_join(thread, 0) != 0) {
        perror("pthread_join");
    }

    tasks.erase(tasks.find(thread));
}

//
// Collect result from finished task.
//
void Application::Collect(const Task *task)
{

}

void Application::Status(std::string message) const
{
    std::ostringstream ss;
    ss << tasks.size() << " active tasks: " << message;
    Output::Info(ss.str());
}

void Application::OnStatusChange(const Task *task) const
{
    if(task->GetStatus() == Task::Starting) {
        Status(task->GetStudent() + " [" + task->GetActivity() + "]" + "\tparams=" + task->ToString());
    } else {
        Status(task->GetStudent() + " [" + task->GetActivity() + "]");
    }
}
